File: src/main/java/de/iteratec/iteraOfficeMap/reservation/ReservationRepository.java
package de.iteratec.iteraOfficeMap.reservation;

import de.iteratec.iteraOfficeMap.workplace.Workplace;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;

public interface ReservationRepository extends JpaRepository<Reservation, Long> {

    List<Reservation> findByStartDateLessThanEqualAndEndDateGreaterThanEqual(LocalDate endDate, LocalDate startDate);

    List<Reservation> findByStartDateLessThanEqualAndEndDateGreaterThanEqualAndWorkplace(
            LocalDate endDate, LocalDate startDate, Workplace workplace);

    Reservation findByStartDateEqualsAndWorkplace(LocalDate startDate, Workplace workplace);

    List<Reservation> findByUsernameEqualsOrderByStartDateAsc(String username);

    Reservation findByWorkplaceAndStartDateEqualsAndEndDateEquals(Workplace workplace, LocalDate startDate, LocalDate endDate);

    @Transactional
    List<Reservation> deleteByWorkplaceAndStartDateAndEndDate(
            Workplace workplace, LocalDate startDate, LocalDate endDate);

}

File: src/main/java/de/iteratec/iteraOfficeMap/reservation/ReservationRepository.kt
package de.iteratec.iteraOfficeMap.reservation

import de.iteratec.iteraOfficeMap.workplace.Workplace
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.transaction.annotation.Transactional

import java.time.LocalDate

interface ReservationRepository : JpaRepository<Reservation, Long> {

    fun findByStartDateLessThanEqualAndEndDateGreaterThanEqual(endDate: LocalDate, startDate: LocalDate): List<Reservation>

    fun findByStartDateLessThanEqualAndEndDateGreaterThanEqualAndWorkplace(
            endDate: LocalDate, startDate: LocalDate, workplace: Workplace): List<Reservation>

    fun findByUsernameEqualsOrderByStartDateAsc(username: String): List<Reservation>

    @Transactional
    fun deleteByWorkplaceAndStartDateAndEndDate(
            workplace: Workplace, startDate: LocalDate, endDate: LocalDate): List<Reservation>

}
